### PR TITLE
trojan: submission

### PR DESCRIPTION
**Pull request by i0ntempest**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.4 19E287
Xcode 11.4.1 11E503a

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by cjones051073** [comment]
Reading the port description i am not sure what to make of this PR. Is it something we really want in macports ?

**Comment by i0ntempest** [comment]
@cjones051073 This is just a proxy client with a weird name, like shadowsocks or v2ray (which is already in macports and homebrew). The description is copied from their github page, maybe I should rewrite it to sound less “offensive”?

**Comment by cjones051073** [comment]
> @cjones051073 This is just a proxy client/server with a weird name, like shadowsocks or v2ray (which is already in macports and homebrew). The description is copied from their github page, maybe I should rewrite it to sound less “offensive”?

I think rephrasing a bit the description would be good, yes, thanks.

File: net/trojan/Portfile [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:filetype=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           cmake 1.1
+PortGroup           github 1.0
+
+name                trojan
+github.setup        trojan-gfw trojan 1.15.1 v
```
Comment (ryandesign) [inline net/trojan/Portfile:8]: You can remove the `name` line. `github.setup` sets it for you.

**Comment by ryandesign** [comment]
How important is MySQL functionality in this port? mysql8 is a big dependency and it is not distributable so if it's not vital I might not enable that variant by default.

When providing optional MySQL functionality in a port, we also usually want to provide multiple variants to let the user choose which implementation they want. For example they might want a different MySQL version, or they might want MariaDB instead of MySQL.

**Comment by i0ntempest** [comment]
Just read the official doc and it can use a MySQL DB for storing user credentials, otherwise would be in the config file. As it's not really mandatory I'll be removing the default variant. Also would you consider providing mysql8, mysql57, mariadb-10.2 variants enough?

**Comment by ryandesign** [comment]
Sure that seems good.

**Comment by i0ntempest** [comment]
Done

File: net/trojan/Portfile [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+checksums           rmd160  b36ac226c6d1cd56c37cd1ef675e0c11fb2d5f19 \
+                    sha256  8627a7cc6e1ebaca93a93cb19e4cc8aac853e1602f117c0470e43db916fded68 \
+                    size    48139
+
+depends_lib-append  port:boost \
+                    path:lib/libssl.dylib:openssl
+
+configure.args-append \
+                    -DCMAKE_INSTALL_SYSCONFDIR=${prefix}/etc -DDEFAULT_CONFIG=${prefix}/etc/trojan/config.json -DENABLE_MYSQL=OFF
```
Comment (mf2k) [inline net/trojan/Portfile:29]: We like to stick to 80 characters maximum width. So please put each argument on its own line.

File: net/trojan/Portfile [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    reinplace "s|/etc/trojan/config.json|${prefix}/etc/trojan/config.json|g" ${worksrcpath}/docs/trojan.1
+}
+
+variant mysql57 conflicts mysql8 mariadb55 mariadb102 description {Enable authencation with a database using MySQL 5.7} {
+    depends_lib-append \
+                        port:mysql57
+    configure.args-replace \
+                        -DENABLE_MYSQL=OFF -DENABLE_MYSQL=ON
+    configure.args-append \
+                        -DMYSQL_INCLUDE_DIR=${prefix}/include/mysql57/mysql/ -DMYSQL_LIBRARIES=${prefix}/lib/mysql57/mysql/
```
Comment (mf2k) [inline net/trojan/Portfile:41]: Put each argument on its own line to stay within 80 characters width.

File: net/trojan/Portfile [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        -L${prefix}/lib/mysql57/mysql -lmysqlclient
+}
+
+variant mysql8 conflicts mysql57 mariadb55 mariadb102 description {Enable authencation with a database using MySQL 8} {
+    depends_lib-append \
+                        port:mysql8
+    configure.args-replace \
+                        -DENABLE_MYSQL=OFF -DENABLE_MYSQL=ON
+    configure.args-append \
+                        -DMYSQL_INCLUDE_DIR=${prefix}/include/mysql8/mysql/ -DMYSQL_LIBRARIES=${prefix}/lib/mysql8/mysql/
```
Comment (mf2k) [inline net/trojan/Portfile:54]: Put each argument on its own line to stay within 80 characters width.

File: net/trojan/Portfile [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                        -L${prefix}/lib/mysql8/mysql -lmysqlclient
+}
+
+variant mariadb55 conflicts mysql57 mysql8 mariadb102 description {Enable authencation with a database using MariaDB 5.5} {
+    depends_lib-append \
+                        port:mariadb
+    configure.args-replace \
+                        -DENABLE_MYSQL=OFF -DENABLE_MYSQL=ON
+    configure.args-append \
+                        -DMYSQL_INCLUDE_DIR=${prefix}/include/mariadb/mysql/ -DMYSQL_LIBRARIES=${prefix}/lib/mariadb/mysql/
```
Comment (mf2k) [inline net/trojan/Portfile:67]: Put each argument on its own line to stay within 80 characters width.

File: net/trojan/Portfile [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                        -L${prefix}/lib/mariadb/mysql -lmysqlclient
+}
+
+variant mariadb102 conflicts mysql57 mysql8 mariadb55 description {Enable authencation with a database using MariaDB 10.2} {
+    depends_lib-append \
+                        port:mariadb-10.2
+    configure.args-replace \
+                        -DENABLE_MYSQL=OFF -DENABLE_MYSQL=ON
+    configure.args-append \
+                        -DMYSQL_INCLUDE_DIR=${prefix}/include/mariadb-10.2/mysql/ -DMYSQL_LIBRARIES=${prefix}/lib/mariadb-10.2/mysql/
```
Comment (mf2k) [inline net/trojan/Portfile:80]: Put each argument on its own line to stay within 80 characters width.

File: net/trojan/Portfile [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                    path:lib/libssl.dylib:openssl
+
+configure.args-append \
+                    -DCMAKE_INSTALL_SYSCONFDIR=${prefix}/etc -DDEFAULT_CONFIG=${prefix}/etc/trojan/config.json -DENABLE_MYSQL=OFF
+
+pre-configure {
+    reinplace "s|/etc/trojan/config.json|${prefix}/etc/trojan/config.json|g" ${worksrcpath}/docs/trojan.1
+}
+
+variant mysql57 conflicts mysql8 mariadb55 mariadb102 description {Enable authencation with a database using MySQL 5.7} {
```
Comment (mf2k) [inline net/trojan/Portfile:35]: This long is also too long. Moving the description to the next line should work.

File: net/trojan/Portfile [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                        -DENABLE_MYSQL=OFF -DENABLE_MYSQL=ON
+    configure.args-append \
+                        -DMYSQL_INCLUDE_DIR=${prefix}/include/mysql57/mysql/ -DMYSQL_LIBRARIES=${prefix}/lib/mysql57/mysql/
+    configure.cppflags-append \
+                        -I${prefix}/include/mysql57/mysql
+    configure.ldflags-append \
+                        -L${prefix}/lib/mysql57/mysql -lmysqlclient
+}
+
+variant mysql8 conflicts mysql57 mariadb55 mariadb102 description {Enable authencation with a database using MySQL 8} {
```
Comment (mf2k) [inline net/trojan/Portfile:48]: Same here, let's move the description to the next line.

File: net/trojan/Portfile [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                        -DENABLE_MYSQL=OFF -DENABLE_MYSQL=ON
+    configure.args-append \
+                        -DMYSQL_INCLUDE_DIR=${prefix}/include/mysql8/mysql/ -DMYSQL_LIBRARIES=${prefix}/lib/mysql8/mysql/
+    configure.cppflags-append \
+                        -I${prefix}/include/mysql8/mysql
+    configure.ldflags-append \
+                        -L${prefix}/lib/mysql8/mysql -lmysqlclient
+}
+
+variant mariadb55 conflicts mysql57 mysql8 mariadb102 description {Enable authencation with a database using MariaDB 5.5} {
```
Comment (mf2k) [inline net/trojan/Portfile:61]: Same here. Move the description to the next line.

**Comment by i0ntempest** [comment]
Done

File: net/trojan/Portfile [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                    -DDEFAULT_CONFIG=${prefix}/etc/trojan/config.json \
+                    -DENABLE_MYSQL=OFF
+
+pre-configure {
+    reinplace "s|/etc/trojan/config.json|${prefix}/etc/trojan/config.json|g" \
+        ${worksrcpath}/docs/trojan.1
+}
+
+variant mysql57 conflicts mysql8 mariadb55 mariadb102 \
+description {Enable authencation with a database using MySQL 5.7} {
```
Comment (cjones051073) [inline net/trojan/Portfile:39]: Typo in description. Its ‘authentication’.

**Comment by i0ntempest** [comment]
Thanks, I'll need to learn the syntax of foreach loop first. Will push later

**Comment by i0ntempest** [comment]
Used a procedure to do the same thing.

File: net/trojan/Portfile [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+notes-append "
+    The default config file is copied to ${prefix}/etc/trojan/config.json and it's content\
+    will be preserved across upgrades and reinstalls. The startup item will use this config\
+    file by default.
+"
+
+startupitem.create          yes
+startupitem.executable      ${prefix}/bin/trojan --config=${prefix}/etc/trojan/config.json
+
+proc dbvariants {dbport prefix} {
```
Comment (cjones051073) [inline net/trojan/Portfile:57]: you can avoid having to pass `prefix` in in each call by just adding the global variable to the scope of the proc. i.e. just add

```
global prefix
```

and then just remove it from the arguments...

Comment (i0ntempest) [reply net/trojan/Portfile]: Thanks

**Pull request merged**